5-AccessControl/2-call-api-groups/API/TodoListAPI/Startup.cs

```diff
@@ -3,65 +3,96 @@
-                        await GraphHelper.GetSignedInUsersGroups(context);
-                        }
-                    };
-                }, options => { Configuration.Bind("AzureAd", options); })
+                           string[] allowedClientApps = { Configuration["AzureAd:ClientId"] }; // In this scenario, client and service share the same clientId
+
+                           string clientappId = context?.Principal?.Claims
+                               .FirstOrDefault(x => x.Type == "azp" || x.Type == "appid")?.Value;
+
+                           if (!allowedClientApps.Contains(clientappId))
+                           {
+                               throw new System.Exception("This client is not authorized");
+                           }
+
+                           // calls method to process groups overage claim.
+                           await GraphHelper.GetSignedInUsersGroups(context);
+
+                           await Task.CompletedTask;
+                        };
+                    }, options => { Configuration.Bind("AzureAd", options); })
                         .EnableTokenAcquisitionToCallDownstreamApi(options => Configuration.Bind("AzureAd", options))
-                        .AddMicrosoftGraph(Configuration.GetSection("GraphBeta"))
-                    .AddInMemoryTokenCaches();
+                        .AddMicrosoftGraph(Configuration.GetSection("MsGraph"))
+                        .AddInMemoryTokenCaches();
 
-            // The following lines code instruct the asp.net core middleware to use the data in the "groups" claim in the Authorize attribute
+            // The following lines code instruct the asp.net core middleware to use the data in the "roles" claim in the Authorize attribute and User.IsInrole()
+            // See https://docs.microsoft.com/aspnet/core/security/authorization/roles for more info.
             services.Configure<JwtBearerOptions>(JwtBearerDefaults.AuthenticationScheme, options =>
             {
-                // The claim in the Jwt token where groups are available.
+                // The claim in the Jwt token where Security Groups are available.
                 options.TokenValidationParameters.RoleClaimType = "groups";
             });
 
             // Adding authorization policies that enforce authorization using Azure AD roles.
             services.AddAuthorization(options =>
             {
-                options.AddPolicy(AuthorizationPolicies.AssignmentToGroupAdminGroupRequired, policy => policy.RequireRole(Configuration["Groups:GroupAdmin"]));
-                options.AddPolicy(AuthorizationPolicies.AssignmentToGroupMemberGroupRequired, policy => policy.RequireRole(Configuration["Groups:GroupMember"]));
+                options.AddPolicy(AuthorizationPolicies.AssignmentToGroupMemberGroupRequired, policy => policy.RequireRole(Configuration["AzureAd:Groups:GroupMember"], Configuration["AzureAd:Groups:GroupAdmin"]));
```

Comment:
>Configuration["AzureAd:Groups:GroupAdmin"] [](http://example.com/codeflow?start=168&length=42)

shouldn't the AssignmentToGroupAdminGroupRequired have both GroupMember and GroupAdmin ids assigned?

Reply:
nah that would let members to access admin only routes